Preliminary implementation of Span-Wagner for finding CO2 density

vegardkv opened the pull request:
Current known shortcomings (intended to be implemented before "un-drafting" this PR):

- 3 tests are failing. The parameters for these tests are at the sublimation or saturation temperature, meaning that density cannot be uniquely determined without additional information. For the same reason, some of the successful tests may be erroneously successful for coincidentally picking the phase corresponding to the tabulated value (when in fact there are two possible outcomes)
- Vectorization. The functions are currently only supporting scalar arguments
- Calculation of Bulk modulus.
- Current line break for these files have been set at 120 characters. From the already existing files, it seems the standard 79 limit is being used

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+    ...     sp.symbols('n1 t1 d1 n2 d2 t2 c2 n3 d3 t3 alpha3 epsilon3 beta3 gamma3 n4 b4 a4 beta4 A4 B4 C4 D4')
+    ... tau, delta = sp.symbols('tau delta', real=True)
+    ... i = sp.symbols('i', integer=True)
+    ... s1 = n1 * delta ** d1 * tau ** t1
```
It seems that if evaluating these expressions over `np.ndarray`s that could be useful for having a bit cleaner code in this module. Regardless, probably moving the sympy part to a separate file, for instance `src/open_petro_elastic/material/span_wagner/equations.py`, seems like a good idea.

vegardkv replied:
Implementation with direct use of sympy for evaluation is now in place.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+
+
+# Constants
+R = 0.1889241 * 1e3  # J / kg K
```
Probably constants should be 
```
CO2_GAS_CONSTANT = 0.1889241e3 # J /Kg K
CO2_CRITICAL_TEMPARATURE = 304.1282 # K
...
```

vegardkv replied:
Fixed.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+"""
+
+
+""" Ideal gas part """
```
Probably, its a good idea to keep the tables in csv files.  If you write the tables into
`src/open_petro_elastic/material/span_wagner/csv_tables/table_31.csv` . And add a file
`src/open_petro_elastic/material/span_wagner/tables.py` containing 
```
import pandas as pd
table_31 = pd.read_scv(
  path.join(path.dirname(__file__), "csv_tables", "table_31.csv")
)
```
Alternatively just keep the table values in a different file.

vegardkv replied:
Moved the coefficients to a separate Python file for now: `src/open_petro_elastic/material/span_wagner/coefficients.py`. Due to varying lengths on the arrays, csv is not very convenient. Other formats are perhaps suited, but for now, having them in a separate Python file at least helps clean up a bit.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+D4 = array([275, 275, 275])
+
+
+def phi(delta, tau, dd, dt):
```
Maybe the name could be `co2_helmholtz_energy` instead of `phi`.

It would be interesting to see if one could implement this has 
```
from .equations import co2_helmholtz_energy
co2_helmholtz_energy_equation_dd = co2_helmholtz_energy.diff(delta)
def co2_helmholtz_energy(delta, tau, dd, dt):
  if dd==1:
    return co2_helmholtz_energy_equation_dd.evaluate(delta, tau)
  ...
```

```
from .equations import residual_helmholtz_energy, ideal_helmholtz_energy
def co2_helmholtz_energy(delta, tau, dd, dt):
  if dd==1:
    return ideal_helmholtz_energy.diff(delta).evaulate(delta, tau) + residual_helmholtz_energy.diff(delta).evaluate(delta, tau)
  ...
```

vegardkv replied:
Fixed.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+    return phi_0(delta, tau, dd, dt) + phi_r(delta, tau, dd, dt)
+
+
+def phi_0(delta, tau, dd, dt):
```
Similarl name change to `ideal_gas_helmholtz_energy`.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+        dd_bigphi = -C4 * (2 * delta - 2) * exp(-C4 * (delta - 1) ** 2 - D4 * (tau - 1) ** 2)
+        dd_s1 = sum(d1 * delta ** (d1 - 1) * n1 * tau ** t1)
+        dd_s2 = sum(-c2 * delta ** (c2 + d2 - 1) * n2 * tau ** t2 * exp(-delta ** c2) + d2 * delta ** (
+                    d2 - 1) * n2 * tau ** t2 * exp(-delta ** c2))
```
Likely some part (or all, or some super-part :P )  of this needs to be made into a look up table, probably something like the following could be cool:

```
from .equations import residual_helmholtz_equation
from scipy import interpolate

residual_helmholtz_dt_dd = interpolate(method="spline", residual_helmholtz_equation.diff(dt, dd).evaluate, ...)
```

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,551 @@
+def carbon_dioxide_bulk_modulus(absolute_temperature, density):
+    """
+    Isentropic bulk modulus, derived from the expression for speed of sound in Table 3 of Span & Wagner
+
```
Some leftover space here.

vegardkv replied:
Fixed.

vegardkv commented:
Just pushed changes based on review feedback. All of the comments are handled, except for implementing a lookup-table approach for expensive calculations, which remain. In addition, a new vectorized density method is implemented.

To summarize the recent changes:
- Moved coefficients to a separate file (helped clean up the code)
- Moved sympy expressions to a separate package (`equations.py`), and changed it so that sympy also evaluates all the expressions (using numpy under the hood).
  - We no longer rely on copy-pasting output based on code in the docstring of `carbon_dioxide.py`, which gives cleaner code.
  - Performance is slightly slower, but comparable (a simple 10,000 random points test was 10% slower).
  - The expressions are currently "lambdified" on import, which causes importing the module to take a 3-5 seconds. With smarter "lambdification" we can probably get away with ~1 second at the cost of more intricate code. If we want even faster, we probably need to pre-calculate the expressions and keep them in a separate file or similar.
- Misc. variable and function name changes
- Implemented a method for co2 density based on `scipy.optimize.newton`, which supports vectorization: `array_carbon_dioxide_density`
  - 10x-15x faster than the `carbon_dioxide_density`, which does not support "proper" vectorization
  - Slower than a look-up table approach according to our experiments
  - Does not depend on look-up table grid resolution and may therefore have some value, even if performance critical code is expected to always use the LUT approach.
  - Contains a few hard-coded "hyper-parameters" that may required tuning for optimal performance. 
  - Falls back to `carbon_dioxide_density` for parameters that fails to converge

JiZur commented:
It looks much leaner, well done :)

JiZur commented on src/open_petro_elastic/material/span_wagner/tables/lookup_table.py:
```diff
@@ -0,0 +1,28 @@
+    grids = [g.flatten() for g in grids2d]
+    z = func(*grids)
+    z_grid = z.reshape(grids2d[0].shape)
+    np.savez(filename, z_grid=z_grid, x=x, y=y)
```
Should we use `savez_compressed` instead of `savez` to store it as compressed file in Github ? Not sure if `np.load()` is able to deal with the compressed npz file directly or if it need un-zipping first

vegardkv replied:
`np.load` supports both, but I am not sure there is anything to gain by using `np.savez_compressed`. It seems the savings are only around 5-10% (for the current settings).

JiZur replied:
sure, then it can probably stay as it is ;)

JiZur reviewed (approved):
Looks good :)

eivindjahren commented on generate_span_wagner_tables.py:
```diff
@@ -0,0 +1,24 @@
+
+out_dir = os.path.join('src', 'open_petro_elastic', 'material', 'span_wagner', 'tables')
+
+t0 = perf_counter()
```
It would be good to have a bit of a refactor of this to be able to integrate this into other tooling later:
```python
import argparse
import numpy as np
from open_petro_elastic.material.span_wagner.tables.lookup_table import generate_lookup_table
from open_petro_elastic.material.span_wagner.carbon_dioxide import carbon_dioxide_density


t_density = 273.15 + np.hstack((np.arange(0, 50, step=0.05), np.arange(50, 200.1, step=0.5)))
p_density = np.hstack((np.arange(0.5, 10, step=0.01), np.arange(10, 100.1, step=0.1)))


def run(args):
    generate_lookup_table(
      lambda _t, _p: carbon_dioxide_density(_t, _p, interpolate=False, force_vapor='auto'),
      t_density, p_density, os.path.join(out_dir, 'carbon_dioxide_density.npz')
    return 0


TOOL_DESCRIPTION = """
used to generate the tables included in open_petro_elastic.material.span_wagner.tables.
"""

def make_parser(prog="generate_sp_tables"):
    ap = argparse.ArgumentParser(prog=prog, description=TOOL_DESCRIPTION)
    ap.add_argument(
        "output_file", type=argparse.FileType("r"), help="The output file to write the density to"
    )
   
    return ap
    
    
def parse_arguments(argv):
    ap = make_parser(argv[0])
    return ap.parse_args(argv[1:])


def main():
    sys.exit(run(parse_arguments(sys.argv)))


if __name__ == "__main__":
    main()

```

vegardkv replied:
Implemented pretty much as suggested with minor modifications.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,420 @@
+        maxiter=10,
+        full_output=True,
+    )
+    # scipy.optimize.newton may not always converge. We need to determine which of the elements of the solution are
```
It would be good to reformat comments so that they do not wrap in git, say 120 character limit.

eivindjahren commented on src/open_petro_elastic/material/span_wagner/carbon_dioxide.py:
```diff
@@ -0,0 +1,420 @@
+
+    :param absolute_temperature: Absolute temperature (K).
+    :param pressure: Pressure (MPa).
+    :param force_vapor: Boolean or 'auto'. If 'auto', the phase of the fluid is automatically determined. However, along
```
Same problem with line length of comments here.
